packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -191,13 +201,29 @@ export class MonacoWorkspace implements lang.Workspace {
         this.textModelService.onDidCreate(model => this.fireDidOpen(model));
     }
 
-    protected _rootUri: string | null = null;
+    protected updateWorkspaceFolders(newRootDirs: FileStat[]): void {
+        const oldWorkspaceUris = this.workspaceFolders.map(folder => folder.uri.toString());
+        const newWorkspaceUris = newRootDirs.map(folder => folder.uri);
+        const added = newWorkspaceUris.filter(uri => oldWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+        const removed = oldWorkspaceUris.filter(uri => newWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+        this._workspaceFolders = newWorkspaceUris.map(this.toWorkspaceFolder);
+        this.onDidChangeWorkspaceFoldersEmitter.fire({ added, removed });
+    }
+
+    protected toWorkspaceFolder(uriString: string, index: number): WorkspaceFolder {
+        return {
+            uri: Uri.parse(uriString),
+            name: uriString.substring(uriString.lastIndexOf('/')) || uriString,
```

Comment:
If you have an encoded URI string here, and you have a file somewhere at `path/to/Foo Bar Folder`, then you will have a URI `file:///path/to/Foo%20Bar%20Folder`. After the `substring` the folder name will be still `Foo%20Bar%20Folder` instead of the desired `Foo Bar Folder`. Please correct me if I am wrong.

Reply:
Probably, something like this would be safer:
```ts
protected toWorkspaceFolder(uriString: string, index: number): WorkspaceFolder {
    const uri = Uri.parse(uriString);
    return {
        uri,
        name: uri.toString(true).substring(uri.toString(true).lastIndexOf('/')) || uri.toString(true),
        index
    };
}
```

Reply:
we have Path api, here it should be base name: `uri.path.base`

Reply:
It's a `vscode` URI here, `uri.path` is a `string`. Did you mean, parse it to a Theia URI and then `path.base`?

Reply:
I did a mixture :-)